src/imitation/util/buffering_wrapper.py

```diff
@@ -0,0 +1,90 @@
+from typing import List
+
+from stable_baselines.common.vec_env import VecEnv, VecEnvWrapper
+
+from imitation.util import rollout
+
+
+class BufferingWrapper(VecEnvWrapper):
+  """Saves transitions of underlying VecEnv.
+
+  Retrieve saved transitions using `pop_transitions()`.
+  """
+
+  def __init__(self, venv: VecEnv, error_on_premature_reset: bool = True):
+    """
+    Args:
+      venv: The wrapped VecEnv.
+      error_on_premature_reset: Error if `reset()` is called on this wrapper
+        and there are saved samples that haven't yet been accessed.
+    """
+    super().__init__(venv)
+    self.error_on_premature_reset = error_on_premature_reset
+    self._trajectories = []
+    self._init_reset = False
+    self._traj_accum = None
+    self._saved_acts = None
+    self.n_transitions = None
+
+  def reset(self, **kwargs):
+    if (self._init_reset and self.error_on_premature_reset
+        and self.n_transitions > 0):  # noqa: E127
+      raise RuntimeError(
+        "BufferingWrapper reset() before samples were accessed")
+    self._init_reset = True
+    self.n_transitions = 0
+    obs = self.venv.reset(**kwargs)
+    self._traj_accum = rollout.TrajectoryAccumulator()
+    for i, ob in enumerate(obs):
+      self._traj_accum.add_step({"obs": ob}, key=i)
+    return obs
+
+  def step_async(self, actions):
+    assert self._init_reset
+    assert self._saved_acts is None
+    self.venv.step_async(actions)
+    self._saved_acts = actions
+
+  def step_wait(self):
+    assert self._init_reset
+    assert self._saved_acts is not None
+    acts, self._saved_acts = self._saved_acts, None
+    obs, rews, dones, infos = self.venv.step_wait()
+    finished_trajs = self._traj_accum.add_steps_and_auto_finish(
+      acts, obs, rews, dones, infos)
+    self._trajectories.extend(finished_trajs)
+    self.n_transitions += self.num_envs
+    return obs, rews, dones, infos
+
+  def _finish_partial_trajectories(self) -> List[rollout.Trajectory]:
+    """Finishes and returns partial trajectories in `self._traj_accum`."""
+    trajs = []
+    for i in range(self.num_envs):
+      # Check that we have any transitions at all.
+      n_transitions = len(self._traj_accum.partial_trajectories[i]) - 1
```

Comment:
Not clear what is going on here. Comment?

Reply:
Ok, added a comment describing why `n_transitions` is calculated this way.